### PR TITLE
Add an option to use Cotire unity build to improve build times

### PR DESCRIPTION
**Pull request by kreuzerkrieg**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Build Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add an option to use Cotire unity build to improve build times

### CODE REVIEW

**Review by alexey-milovidov — APPROVED** [review]

**Comment by filimonov** [comment]
Can you share the results / observations about build time with and w/o cotire?

**Comment by filimonov** [comment]
Also cotire readme says that
> The functionality provided by cotire has been superseded by features added to CMake 3.16. Support for pre-compiling and unity builds is now built into CMake. Thus, there will not be any further updates or support for this project.

Would be interesting to compare with option "use latest cmake"

**Comment by alexey-milovidov** [comment]
Switch to the latest CMake is possible (and appreciated) if we will add as simple as possible way to install it to the build instruction.

**Comment by kreuzerkrieg** [comment]
> Switch to the latest CMake is possible (and appreciated) if we will add as simple as possible way to install it to the build instruction.

Later on I will test it with new cmake
Meanwhile, what happens to the build? looks like it is stuck

**Comment by alexey-milovidov** [comment]
Build is not even started, something was stuck in CI.

File: CMakeLists.txt [hunk 1]
```diff
@@ -14,6 +14,12 @@ endforeach()
 
 project(ClickHouse)
 
+option(ENABLE_COTIRE "Enable cotire unity build" ON)
```
Comment (alexey-milovidov) [inline CMakeLists.txt]: Should we check that submodule exists?

Comment (kreuzerkrieg) [reply CMakeLists.txt]: dunno... is this what done with the rest of submodules? if so, I will check how it was done with other submodules

**Comment by alexey-milovidov** [comment]
There is some incompatibility, please look at the build log: https://clickhouse-builds.s3.yandex.net/8894/f34b341427daa2fdeffabe2839a7f61cbb8394c9/build_log_599049204_1580382774.txt

File: CMakeLists.txt (hunk 1 above)
Comment (alexey-milovidov) [reply CMakeLists.txt]: Yes, we still have limited support for build without submodules.

**Comment by kreuzerkrieg** [comment]
> There is some incompatibility, please look at the build log: https://clickhouse-builds.s3.yandex.net/8894/f34b341427daa2fdeffabe2839a7f61cbb8394c9/build_log_599049204_1580382774.txt

yes, I know, yesterday when it finaly started to build I saw this problem
still checking.

**Comment by kreuzerkrieg** [comment]
Ok, I have to admit two things.
First I've tested it on my branch stemming from the `stable` instead of testing on master.
Second, I have no idea what to do with it. I've found the root cause, it is the `add_cxx_compile_options` in cmake, more specific - TARGET_PROPERTY:LINKER_LANGUAGE expression generator. when it was easy to substitute it, it led to a horde of other compilation problems in PCH files generated by `cotire`
I think, since the `cotire` isnt supported anymore by the author because of the CMake3.16.3  there is no sense to try and make it work. Instead I checked more feasible and promising approach - use latest cmake and check if we can benefit from its new featured - PCHing and unity build. Unity build looks very promising, but the ODR hell broke loose and I spent a day trying to fix it. We will see if I can complete this task in reasonable time. PCHing looks easier to achieve, not sure how much it will cut in build time. Thats all for today. Sorry for inconvenience. @alexey-milovidov if you feel like closing this PR, go ahead

**Comment by alexey-milovidov** [comment]
Ok.

**Pull request closed without merging**